File: app/Services/ACOService.php
<?php

namespace App\Services;

class ACOService
{
    private int $numAnts;
    private int $maxIter;
    private float $alpha;
    private float $beta;
    private float $rho;
    private float $q0;
    private float $Q;
    private float $initPheromone;
    private array $distMatrix;

    public function __construct(
        array $distMatrix,
        int   $numAnts = 10,
        int   $maxIter = 50,
        float $alpha   = 1.0,
        float $beta    = 2.0,
        float $rho     = 0.5,
        float $q0      = 0.7,
        float $Q       = 100,
        float $initPheromone = 0.1
    ) {
        $this->distMatrix    = $distMatrix;
        $this->numAnts       = $numAnts;
        $this->maxIter       = $maxIter;
        $this->alpha         = $alpha;
        $this->beta          = $beta;
        $this->rho           = $rho;
        $this->q0            = $q0;
        $this->Q             = $Q;
        $this->initPheromone = $initPheromone;
    }

    /**
     * Menyelesaikan TSP untuk kota-kota yang harus dikunjungi
     * dimulai dari startCity. Mengembalikan rute optimal + jarak + histori iterasi.
     */
    public function solve(string $startCity, array $citiesToVisit): array 
{
    if (empty($citiesToVisit)) {
        return ['route' => [$startCity], 'distance' => 0, 'iterations' => []];
    }
    $citiesToVisit = array_values(array_unique($citiesToVisit));

    $allCities = array_unique(array_merge([$startCity], $citiesToVisit));
    $n = count($allCities);

    // Inisialisasi matriks feromon
    $pheromone = [];
    foreach ($allCities as $ci) {
        foreach ($allCities as $cj) {
            if ($ci !== $cj) {
                $pheromone[$ci][$cj] = $this->initPheromone;
            }
        }
    }

    $bestRoute    = null;
    $bestDistance  = INF;
    $iterHistory  = [];

    for ($iter = 0; $iter < $this->maxIter; $iter++) {
        $antRoutes     = [];
        $antDistances  = [];

        for ($a = 0; $a < $this->numAnts; $a++) {
            $route    = $this->buildRoute($startCity, $citiesToVisit, $pheromone);
            $distance = $this->routeDistance($route);
            $antRoutes[]    = $route;
            $antDistances[] = $distance;

            if ($distance < $bestDistance) {
                $bestDistance = $distance;
                $bestRoute    = $route;
            }
        }

        // Evaporasi feromon
        foreach ($pheromone as $ci => &$row) {
            foreach ($row as $cj => &$p) {
                $p *= (1 - $this->rho);
                if ($p < 1e-10) $p = 1e-10;
            }
        }
        unset($row, $p);

        // Deposisi feromon
        for ($a = 0; $a < $this->numAnts; $a++) {
            $deposit = $this->Q / max($antDistances[$a], 1);
            $route = $antRoutes[$a];
            for ($k = 0; $k < count($route) - 1; $k++) {
                if (isset($pheromone[$route[$k]][$route[$k + 1]])) {
                    $pheromone[$route[$k]][$route[$k + 1]] += $deposit;
                }
            }
        }

        $iterHistory[] = [
            'iteration'     => $iter + 1,
            'best_distance' => round($bestDistance, 2),
            'avg_distance'  => round(array_sum($antDistances) / count($antDistances), 2),
        ];
    }

    return [
        'route'      => $bestRoute,
        'distance'   => round($bestDistance, 2),
        'iterations' => $iterHistory
    ];
}

    private function buildRoute(string $start, array $toVisit, array &$phero): array
    {
        $route     = [$start];
        $unvisited = $toVisit;
        $current   = $start;

        while (!empty($unvisited)) {
            $next       = $this->pickNext($current, $unvisited, $phero);
            $route[]    = $next;
            $unvisited  = array_values(array_diff($unvisited, [$next]));
            $current    = $next;
        }
        return $route;
    }

    private function pickNext(string $current, array $unvisited, array &$phero): string
    {
        // Exploitasi vs Eksplorasi
        if (mt_rand() / mt_getrandmax() < $this->q0) {
            $bestCity = $unvisited[0];
            $bestVal  = -1;
            foreach ($unvisited as $city) {
                $tau = $phero[$current][$city] ?? $this->initPheromone;
                $eta = 1.0 / max($this->distMatrix[$current][$city] ?? 1, 1);
                $val = pow($tau, $this->alpha) * pow($eta, $this->beta);
                if ($val > $bestVal) {
                    $bestVal  = $val;
                    $bestCity = $city;
                }
            }
            return $bestCity;
        }

        // Roulette wheel
        $probs = [];
        $total = 0;
        foreach ($unvisited as $city) {
            $tau = $phero[$current][$city] ?? $this->initPheromone;
            $eta = 1.0 / max($this->distMatrix[$current][$city] ?? 1, 1);
            $p   = pow($tau, $this->alpha) * pow($eta, $this->beta);
            $probs[$city] = $p;
            $total += $p;
        }

        $r = mt_rand() / mt_getrandmax() * $total;
        $cum = 0;
        foreach ($unvisited as $city) {
            $cum += $probs[$city];
            if ($cum >= $r) return $city;
        }
        return $unvisited[count($unvisited) - 1];
    }

    private function routeDistance(array $route): float
    {
        $total = 0;
        for ($i = 0; $i < count($route) - 1; $i++) {
            $total += $this->distMatrix[$route[$i]][$route[$i + 1]] ?? INF;
        }
        return $total;
    }
}

File: services/aco_service.py
import random
import math

class ACOService:
    def __init__(self, dist_matrix, num_ants=10, max_iter=50, alpha=1.0, beta=2.0, rho=0.5, q0=0.7, Q=100, init_pheromone=0.1):
        self.dist_matrix = dist_matrix
        self.num_ants = num_ants
        self.max_iter = max_iter
        self.alpha = alpha
        self.beta = beta
        self.rho = rho
        self.q0 = q0
        self.Q = Q
        self.init_pheromone = init_pheromone

    def solve(self, start_city, cities_to_visit, return_to_start=True):
        """
        Mencari rute terbaik dari start_city mengunjungi semua cities_to_visit.
        Jika return_to_start=True (default), truk akan kembali ke kota asal
        (gudang) di akhir rute, dan jarak tempuh tersebut ikut dihitung dalam
        total_distance. Ini merepresentasikan siklus pengiriman yang nyata:
        truk berangkat dari gudang, mengantar semua barang, lalu pulang ke
        gudang yang sama. Tanpa ini, biaya BBM yang dihitung akan lebih
        rendah dari kenyataan karena jarak pulang tidak diperhitungkan.
        """
        if not cities_to_visit:
            return {'route': [start_city], 'distance': 0, 'iterations': []}
            
        cities_to_visit = list(dict.fromkeys(cities_to_visit)) # Unique
        
        all_cities = list(dict.fromkeys([start_city] + cities_to_visit))
        
        # Init pheromone
        pheromone = {}
        for ci in all_cities:
            pheromone[ci] = {}
            for cj in all_cities:
                if ci != cj:
                    pheromone[ci][cj] = self.init_pheromone
                    
        best_route = None
        best_distance = float('inf')
        iter_history = []
        
        for iteration in range(self.max_iter):
            ant_routes = []
            ant_distances = []
            
            for a in range(self.num_ants):
                route = self._build_route(start_city, cities_to_visit, pheromone)
                if return_to_start:
                    route = route + [start_city]
                distance = self._route_distance(route)
                ant_routes.append(route)
                ant_distances.append(distance)
                
                if distance < best_distance:
                    best_distance = distance
                    best_route = route
                    
            # Evaporasi feromon
            for ci in pheromone:
                for cj in pheromone[ci]:
                    pheromone[ci][cj] *= (1 - self.rho)
                    if pheromone[ci][cj] < 1e-10:
                        pheromone[ci][cj] = 1e-10
                        
            # Deposisi feromon (termasuk edge pulang ke gudang, jika
            # return_to_start aktif, sehingga ACO ikut "belajar" memilih
            # kota terakhir yang dekat dengan gudang demi meminimalkan
            # total jarak round-trip, bukan hanya jarak berangkat)
            for a in range(self.num_ants):
                deposit = self.Q / max(ant_distances[a], 1)
                route = ant_routes[a]
                for k in range(len(route) - 1):
                    ci, cj = route[k], route[k+1]
                    if ci in pheromone and cj in pheromone[ci]:
                        pheromone[ci][cj] += deposit
                        
            iter_history.append({
                'iteration': iteration + 1,
                'best_distance': round(best_distance, 2),
                'avg_distance': round(sum(ant_distances) / len(ant_distances), 2)
            })
            
        return {
            'route': best_route,
            'distance': round(best_distance, 2),
            'iterations': iter_history
        }

    def _build_route(self, start, to_visit, phero):
        route = [start]
        unvisited = to_visit[:]
        current = start
        
        while unvisited:
            next_city = self._pick_next(current, unvisited, phero)
            route.append(next_city)
            unvisited.remove(next_city)
            current = next_city
            
        return route

    def _pick_next(self, current, unvisited, phero):
        # Exploitasi
        if random.random() < self.q0:
            best_city = unvisited[0]
            best_val = -1
            for city in unvisited:
                tau = phero.get(current, {}).get(city, self.init_pheromone)
                eta = 1.0 / max(self.dist_matrix.get(current, {}).get(city, 1), 1)
                val = (tau ** self.alpha) * (eta ** self.beta)
                if val > best_val:
                    best_val = val
                    best_city = city
            return best_city
            
        # Eksplorasi (Roulette wheel)
        probs = {}
        total = 0
        for city in unvisited:
            tau = phero.get(current, {}).get(city, self.init_pheromone)
            eta = 1.0 / max(self.dist_matrix.get(current, {}).get(city, 1), 1)
            p = (tau ** self.alpha) * (eta ** self.beta)
            probs[city] = p
            total += p
            
        r = random.random() * total
        cum = 0
        for city in unvisited:
            cum += probs[city]
            if cum >= r:
                return city
        return unvisited[-1]

    def _route_distance(self, route):
        total = 0
        for i in range(len(route) - 1):
            total += self.dist_matrix.get(route[i], {}).get(route[i+1], float('inf'))
        return total